### ISSUE TITLE
[Bug] when addField is used to add from a relationship it's very slow

### ISSUE THREAD

**Issue by realtebo** [Laravel-Backpack/CRUD#4618]
# Bug report

### What I did

I've a PracticeCrudController where, in editing, I must directly edit as text the practice -> customer -> name. 

I added this in `setupUpdateOperation` of `PracticeCrudController`

        $this->crud->addField([
            'label' => 'Partita iva',
            'type' => 'text',
            'tab' => 'Cliente/Richiedente',
            'name' => 'customer.vat_registration_number',
            'wrapper' => ['class' => 'form-group col-md-6'],
            'attributes' => [
                'readonly' => true
            ]
        ]);

Of course we customized update method to handle this scenario. But the problem happens when we LOAD the /edit route, not  (not only) when saving

Practice is in a 1:1 relation with Customer 

Practice model has 

    public function customer()
    {
        return $this->hasOne(Customer::class, 'practice_id');
    }

then the customers.practice_id is already a foreign key (so it's indexed) pointing to practice.jd

Note: It CANNOT be a data amount problem because we have 1 row only in each table.

Also, this happens in 2 different windows machines with 2 different mariadb version

### What I expected to happen

It works and no big difference on timing 

### What happened

Adding this brings timing (recorderd using debug bar)  of our page from this

![image](https://user-images.githubusercontent.com/1268699/186389409-de108fe6-d38f-4f55-81b8-2ebffa767950.png)

To this 

![image](https://user-images.githubusercontent.com/1268699/186389559-16a48c97-898c-4990-bf90-dff28e02cffa.png)

### What I've already tried to fix it

I  am unable to fix But I tried to use debug bar time-traking system to undestand where is the bottleneck

From `$this->crud->addField` I indentified the 'slow path' as 
- vendor/backpack/crud/src/app/Library/CrudPanel/Traits/Fields.php -> addField 
- vendor/backpack/crud/src/app/Library/CrudPanel/Traits/Fields.php -> makeSureFieldHasNecessaryAttributes 
- vendor/backpack/crud/src/app/Library/CrudPanel/Traits/Fields.php -> makeSureFieldHasRelationshipAttributes
- vendor/backpack/crud/src/app/Library/CrudPanel/Traits/FieldsProtectedMethods.php -> makeSureFieldHasAttribute
- vendor/backpack/crud/src/app/Models/Traits/HasIdentifiableAttribute.php -> identifiableAttribute
- vendor/backpack/crud/src/app/Models/Traits/HasIdentifiableAttribute.php -> guessIdentifiableColumnName
  -  **Slow code is the call to**
     $table = $instance->getTableWithPrefix();
     and
     $columns = $conn->getDoctrineSchemaManager()->listTableColumns($table);

This is my time-tracking code, I modified directly the code on 

    protected function makeSureFieldHasAttribute($field)
    {

        // if there's a model defined, but no attribute
        // guess an attribute using the identifiableAttribute functionality in CrudTrait
        if (isset($field['model']) && ! isset($field['attribute']) && method_exists($field['model'], 'identifiableAttribute')) {

            \Debugbar::startMeasure('makeSureFieldHasAttribute -> call_user_func');
            $field['attribute'] = call_user_func([(new $field['model']()), 'identifiableAttribute']);
            \Debugbar::stopMeasure('makeSureFieldHasAttribute -> call_user_func');
        }

        return $field;
    }

### Is it a bug in the latest version of Backpack?

After I run ```composer update backpack/crud``` the bug... is it still there?

Yes,
Also, I _downgraded_ from actual 5.35 to 5.00 and nothing changes

### Backpack, Laravel, PHP, DB version

When I run ```php artisan backpack:version``` the output is:
```
### PHP VERSION:                                                                                     
PHP 8.1.2 (cli) (built: Jan 19 2022 10:13:52) (NTS Visual C++ 2019 x64)                              
Copyright (c) The PHP Group                                                                          
Zend Engine v4.1.2, Copyright (c) Zend Technologies                                                  
                                                                                                     
### LARAVEL VERSION:                                                                                 
v9.26.0@53e20ba7c4df1d1628102811f5b7119e138247f8                                                     
                                                                                                     
### BACKPACK VERSION:                                                                                
5.3.5@0f1f05f917db8da7f3ae415bbada8474095b85df                                                       
                                                                                                    
```

**Comment by realtebo** [comment]
Further investigations
-

Modified code as

    protected function makeSureFieldHasAttribute($field)
    {

        // if there's a model defined, but no attribute
        // guess an attribute using the identifiableAttribute functionality in CrudTrait
        if (isset($field['model']) && ! isset($field['attribute']) && method_exists($field['model'], 'identifiableAttribute')) {

            \Debugbar::startMeasure('makeSureFieldHasAttribute -> new instance');
            $instance = new $field['model'];
            \Debugbar::stopMeasure('makeSureFieldHasAttribute -> new instance');

            \Debugbar::startMeasure('makeSureFieldHasAttribute -> execution');
            $execution = (new $field['model']());
            \Debugbar::stopMeasure('makeSureFieldHasAttribute -> execution');

            \Debugbar::startMeasure('makeSureFieldHasAttribute -> call_user_func');
            $field['attribute'] = call_user_func([$execution, 'identifiableAttribute']);
            \Debugbar::stopMeasure('makeSureFieldHasAttribute -> call_user_func');
        }

        return $field;
    }


Timings

![image](https://user-images.githubusercontent.com/1268699/186392879-1cd5cb3b-49dc-4746-a1e5-1025d012b345.png)

I did this to _exclude a problem into my model class_

**Comment by realtebo** [comment]
Customer Class
-

I see nothing of strange in this class

```
use Illuminate\Database\Eloquent\Model;
use Backpack\CRUD\app\Models\Traits\CrudTrait;
...
class Customer extends Model
{
    use UseUuidsAsPrimaryKeyTrait;
    use CrudTrait;
}
```

```

We have _not_ customized the constructor.
This class has only relationships and nothing else, no functions, no scopes, ...

The trait for using uuid as primary key is here

```
<?php

namespace App\Traits;

use Illuminate\Support\Str;

trait UseUuidsAsPrimaryKeyTrait
{
    /**
     * Boot function from Laravel.
     */
    protected static function boot()
    {
        parent::boot();
        static::creating(function ($model) {
            if (empty($model->{$model->getKeyName()})) {
                $model->{$model->getKeyName()} = Str::uuid()->toString();
            }
        });
    }

    /**
     * Get the value indicating whether the IDs are incrementing.
     *
     * @return bool
     */
    public function getIncrementing()
    {
        return false;
    }

    /**
     * Get the auto-incrementing key type.
     *
     * @return string
     */
    public function getKeyType()
    {
        return 'string';
    }
}
```

**Comment by realtebo** [comment]
Workaround
-

I 'fixed' the bad behaviour changing 

        $this->crud->addField([
            'label' => 'Denominazione',
            'type' => 'text',
            'tab' => 'Cliente/Richiedente',
            'name' => 'customer.name',
            'wrapper' => ['class' => 'form-group col-md-6'],
            'attributes' => [
                'class' => 'form-control text-uppercase'
            ]
        ]);

into 

        $this->crud->addField([
            'label' => 'Denominazione',
            'type' => 'text',
            'tab' => 'Cliente/Richiedente',
            'name' => 'customer.name',
            'attribute' => 'name',
            'wrapper' => ['class' => 'form-group col-md-6'],
            'attributes' => [
                'class' => 'form-control text-uppercase'
            ]
        ]);

Note: I simply added 'attribute'

After seeing this piece of code


    protected function makeSureFieldHasAttribute($field)
    {
        \Log::info("makeSureFieldHasAttribute (field)", [ "field" => $field ]);
        // if there's a model defined, but no attribute
        // guess an attribute using the identifiableAttribute functionality in CrudTrait
        if (isset($field['model']) && ! isset($field['attribute']) && method_exists($field['model'], 'identifiableAttribute')) {

            // slow
            $field['attribute'] = call_user_func([(new $field['model']()), 'identifiableAttribute']);
            // slow
        }

        return $field;
    }

So I concluded that adding attribute will not invoke the guess chain

But is still a mystery why this takes 7 seconds

        \Debugbar::startMeasure('guessIdentifiableColumnName -> getTableWithPrefix');
        $table = $instance->getTableWithPrefix();
        \Debugbar::stoptMeasure('guessIdentifiableColumnName -> getTableWithPrefix');

when inside execution takes only few of microseconds !

    public function getTableWithPrefix()
    {
        \Debugbar::startMeasure('getTableWithPrefix -> inner');
        $prefix = $this->getConnection()->getTablePrefix();
        $tableName = $this->getTable();
        $out = $prefix.$tableName;
        \Debugbar::stopMeasure('getTableWithPrefix -> inner');
        \Log::info ($out);
        return $out;
    }

See this

![image](https://user-images.githubusercontent.com/1268699/186404039-b80f8df2-dc1b-414c-a55f-77500e193692.png)

**Comment by realtebo** [comment]
Another workaround
-

Adding to `Customers.php` model class this

   public function identifiableAttribute()
    {
        return "name";
    }

This will avoid calling internal guess chain

**Comment by pxpm** [comment]
Thank you @realtebo amazing issue description. 🙏 

I will have a look at it tomorrow, already assigned myself. 

Cheers

**Comment by pxpm** [comment]
Hello @realtebo 
Any chance the cause here are the `uuid` as primary keys ?

Any chance you can reproduce this in a model with regular "primary incrementing id" ? 

I've been trying some solutions to see if I can spot any difference, the only difference I see now is the fact that you use UUID and not incrementing ids. 

Let me know if you have other models around to test or I should create my own uuid model and try it out. 

I am also using mysql, not MariaDB ... I will run my tests on a maria db instance that I am setting up and let you know if that's the culprite, but I kinda doubt it ... but who knows!

Cheers

**Comment by realtebo** [comment]
Sorry, entire project, a big one, and this is the biggest one, is made using uuid for a customer request to avoid 'simple' ids on get urls.

I already tried against a mysql instead of mariadb. nothing changes.

the ONLY thing I can theorize is that PHP itself is "panicing" because of "too many" levels of inheritance, too may  traits, or so on.

I am sure it's not a bug itself but a very strange edge-case. Nothing similar ever happined to us, and we make only big strange and more strange projects ... sorry. 

> a suggestion, because related:  in these cases I used syntax: "relation.attribute". The code is able to understand that using a dot means i am usign a relation, so I think that _before_ try to guess identifable attribute, you should first check if 'attribute' is a db field. Why try to guess something we already specified?

**Comment by pxpm** [comment]
Hello @realtebo 

following the answer to your question, this is my reply: https://github.com/Laravel-Backpack/CRUD/pull/4621 👍 
Hope it solves the issue here.
I couldn't reproduce it, but it seems we were doing more work than needed, and we had already implemented this solution for columns too. 

If you happen to give that PR a go, let me know if it works for you. 

Cheers

**Comment by jorgetwgroup** [comment]
Hello @realtebo 

Could you check with PR https://github.com/Laravel-Backpack/CRUD/pull/4621 and let us know about your results.

Did you say are using Windows and MySQL or MariaDB, did you run this over Docker or what kind of server?

Thanks in advance.

Cheers.

**Comment by pxpm** [comment]
@realtebo I am comming back here because this issue bit me in the *ss today! 
Thank god you opened this issue, and I didn't start ripping my development environment apart before trying to figure out the root cause for such high loading times. 

I got **20s** loading time on Monster Crud in demo ... out of the blue! 🙃 

I started by testing possible bootlenecks that we might had introduced recently because at the time you opened this issue, as I told you, I was not able to reproduce it. 🤷‍♂️ 

I tried the PR we created from this discussion ... it helped a bit, but not much. 

I kept going and endup with the last possible solution, it must be something to do with DB engine. And yes, it does have **EVERYTHING** to do with db engine. 

Just by updating mysql version from 5.7 to 8.0 the page went from 20s to 800ms load time. Even without the other branch we created. 
I dig a bit on your use case and found out that MariaDB was started from version 5.x so it may have the same problem as mysql 5.7. 

We are working on a solution for 5.7, it wouldn't be as fast, but it can get the monster crud back to 2s load time 🎉 🎉 🎉 🎉 

The reason why I didn't get it before and now I got it ... still unknown! 🤷‍♂️ 

Cheers
